design-proposals/2022-10-27-rendering-modularization.md

```diff
@@ -0,0 +1,495 @@
+
+We can make things like the location puck simpler.  If we can expose much of this outside the toolkit we may even make it possible to do things like your own location puck, or similar functionality.
+
+By moving to this approach we can let the rendering logic deal only with its own classes.  Rather than letting implementaton propagate all the way down to the render_*_layer, which has to very specifically implement them, we can make layers ask for what they need and let the SDK level implement it however it may feel like.  This is particularly important for Metal's indirect mode.
+
+### Atlases: Line, Image, Pattern, and Glyph
+
+_Addresses core functionality [#3](#core)._
+
+Having a lot of little textures floating around is inefficient.  This was very much the case in OpenGL, but it's still true in Metal and other SDKs.  There's a long conversation to be had about that, but it's somewhat moot.  MapLibre Native uses atlases, so we need to consider them.
+
+All of these atlases are essentially Texture Atlases.  That is, a group of individual textures gathered together for efficiency.  Shaders reference these with a little offset into a larger texture.  The shaders may not even be a aware of it, I haven't checked.
+
+#### The way it is now
+
+Right now each of the atlas types has a reference within the render loop (discussed below).  They also seem to do their image uploading on the rendering thread.  That's unnecessary even in OpenGL, but it is simple.
+
+#### Required Changes
+
+The rendering loop shouldn't need to know about any of the atlases. That bookkeeping is better done as a callback.  That would open up the possibility of doing more atlas types in the future.
+
+Further, the atlases should try to do their data uploading on other threads.  If you're doing this with OpenGL you need to open shared contexts and admittedly, it can get a bit complex.  With Metal it's much simpler.
+
+Much of this could be done with expanded observers as discussed below.
+
+#### Benefits
+
+Less work done in the rendering loop is always better.  More CPU available for UI developers to update the curves on their buttons.
+
+Making the atlas updates more generic might allow for new types, or even the fabled Dynamic Texture Atlas.  We have one in WhirlyGlobe-Maply and it's great for adding random geometry and textures into the mix.  But that's outside the scope of this proposal.
+
+### Main Render Loops
+
+_Addresses refactoring that make migration easier, particularly our [north stars](#north-stars)._
+
+Every real time rendering toolkit has a main rendering loop or something like it.  It's not actually a loop anymore, it's a callback.  Years ago there were big fights between the regular UI people and the real time rendering people and the first group won.  But we have our dignity, so it's a "rendering loop".
```

Comment:
> But we have our dignity, so it's a "rendering loop".

I don't understand this statement, can you clarify this paragraph?

Reply:
This is largely historical.  In real time rendering we used to actually have a rendering loop.  main() would call down into a loop which drew and then waited for a set period of time for the next frame and then drew again.  I realized that we still refer to that as a "rendering loop" even though it's just callbacks from the system now and has been for quite some time.

Reply:
I've rewritten this part.